### DISCUSSION TITLE
Conservation of energy doesn't hold with joint constrained bodies?

### DISCUSSION THREAD

**Discussion by LordXerus** [dyn4j/dyn4j#226] [Q&A]
I thought I would be able to make a game of Pong with this physic engine. However, the ball would slow down horizontally every time it hits the paddles.

The paddles are `MassType.NORMAL` bodies with a prismatic joint to a fixture-less `MassType.INFINITE` body. It seems that the horizontal paddles are absorbing energy from the ball, but then losing the energy due to the joint.

Is this expected behavior? and if so how to prevent the ball from losing energy?

**Comment by wnbittle** [comment]
@LordXerus  are you setting the restitution of the fixtures?  The default restitution is 0, so I'm guessing you are, but wanted to check.

In general, there is a loss of energy of the simulation over time - this is by design.  Systems with conservation of energy/momentum are very hard, if not impossible (I'm looking at you floating point), and systems that increase total energy will eventually explode.  Is the joint absorbing the energy?  It's possible.  The system is being solved iteratively, converging to a solution, rather than solving the entire system directly.  If you have an example of the situation that you can share I can take a further look.

That said, there are ways to increase the energy of the system or compensate for the loss of energy.  For example, you can have higher restitution values than 1 (simulating paddles like you'd find on a pin ball machine).

**Reply by LordXerus** [reply]
I've looked at the `CoefficientMixer` class, which I believe determines the final restitution in the collision. From my understanding, the restitution is the larger of the two, so what I have done is set the restitution of the ball (which bounces against everything) to `1.0`.

I want to recreate a Pong-like simulation; the ball should roughly keep the same speed, and only change in direction. However, after it hits the paddle, the horizontal velocity of the ball roughly halves. I think this is because the `1.0` restitution splits the energy to both the ball and the paddle evenly; however, the energy in the paddle is immediately lost to the prismatic joint. 

I guess to mitigate this I can to set paddles' restitution to `2.0` but at the same time I don't really want the player controlled bodies to bounce against the walls (which I'm using to keep the player and balls in boundary. 

Here's the project I'm working on: [Pong](https://github.com/LordXerus/physics-pong)


Here's a compiled release: [Java_8.zip](https://github.com/dyn4j/dyn4j/files/8984121/Java_8.zip)

**Comment by wnbittle** [comment]
Briefly looking at the code I thought of two things which could be happening:

When you render the ball are you rendering it's rotation?  If there's friction on the ball/paddle, it's possible part of the linear velocity is converted to angular velocity.  You won't see it if the ball is rendered as a solid circle.  One way to eliminate this possibility is to set the mass of the ball to MassType.FIXED_ANGULAR_VELOCITY.  This gives the body infinite angular mass, meaning that it will rotate at a fixed linear velocity - typically zero - i.e. never rotating.

I noticed that the density of the ball is 1.0 and the density of the paddle is 0.1.  The problem with this is that the majority of the energy from the ball will be transferred to the paddle because the paddle is the lighter object.  The joint then sees that the paddle is violating the constraint, and removes the energy to keep it on the track.
